### PR TITLE
Add online DQM plots for Run 3 L1T shower trigger

### PR DESCRIPTION
**Pull request by zuoxunwu**
#### PR description:

Added L1T DQM module for CSC local shower and EMTF regional shower in the Run 3 trigger.

This DQM module is complete and tested with Run 3 MC samples.
As of CMSSW_12_2_0_pre2, there is an unpacker for CSC showers ([PR#35461](https://github.com/cms-sw/cmssw/pull/35461))but not for EMTF showers. Therefore at the moment the shower DQM cannot be used in data-taking and is not enabled in the online DQM sequence. It will be enabled once the EMTF shower unpacker is ready.
In this PR we do not expect any change in the output.

#### PR validation:

The DQM plots are tested with HTo2LongLivedTo4b samples and Neutrino gun samples. 
Passed the unit test and matrix test as there should be no change in the output.
Code format changes are applied.

@dildick may want to watch this.

### CODE REVIEW

**Comment by jfernan2** [comment]
@zuoxunwu 
Before moving onward, I have several questions/requests for you:
- please add yourself along with your github username in the comments section to the corresponding e-group in: https://twiki.cern.ch/twiki/bin/viewauth/CMS/DQMContacts#L1T
- I understand from the title of the PR that it is adding plots to ONline DQM, however the PR is in master and Online DQM is running on 12_0_X at present. Do you foresee to use this in current MWGR data taking? If so, we need a backport, but I understand unpacker is not ready yet
- You have tested the PR with MC, but if the plots to appear (once the unpacker is in) are for Online, it would eb nice to have some check on data and with some upload to a private Online DQM GUI to validate them 

Thanks

**Comment by jfernan2** [comment]
please test

**Comment by zuoxunwu** [comment]
@jfernan2 
I have run the shower DQM on a zero-bias test run (346512) and viewed it on a private DQM GUI. Everything worked well. 
Most DQM plots for this run are empty, except the bins used for event counting. (This run is low energy and low pileup and we do not expect showers in muon chambers.) I am attaching a screenshot of my private DQM GUI.
Screen Shot 2021-11-17 at 6 41 01 PM

Just as a reference, the plots on MC samples can be found in [these slides](https://indico.cern.ch/event/1097365/#3-dqm).

Overall I think the shower DQM module works as intended and this PR can be merged. Let me know if you have further comments. Thanks!

**Comment by jfernan2** [comment]
Thank you very much @zuoxunwu !

**Comment by jfernan2** [comment]
+1

File: DQM/L1TMonitor/src/L1TStage2Shower.cc [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                                                           {{1, 3}, 6},
+                                                           {{2, 1}, 5},
+                                                           {{2, 2}, 4},
+                                                           {{3, 1}, 3},
+                                                           {{3, 2}, 2},
+                                                           {{4, 1}, 1},
+                                                           {{4, 2}, 0}};
+
+  // Fill CSC local shower plots
+  for (auto element = CscShowers->begin(); element != CscShowers->end(); ++element) {
```
Comment (perrotta) [inline DQM/L1TMonitor/src/L1TStage2Shower.cc:96]:
```suggestion
  for (auto const& element : CscShowers) {
```

File: DQM/L1TMonitor/src/L1TStage2Shower.cc [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        if (cscShower->isNominalInTime())
+          cscShowerChamber->Fill(chamber, 2.5);
+        if (cscShower->isTightInTime())
+          cscShowerChamber->Fill(chamber, 3.5);
+      }
+    }
+  }
+
+  // Fill EMTF regional shower plots
+  for (auto Shower = EmtfShowers->begin(); Shower != EmtfShowers->end(); ++Shower) {
```
Comment (perrotta) [inline DQM/L1TMonitor/src/L1TStage2Shower.cc:168]:
```suggestion
  for (auto const& shower : EmtfShowers) {
```
(lowercase "s" is more in line with the rest of the code)

File: DQM/L1TMonitor/src/L1TStage2Shower.cc [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    int endcap = Shower->endcap();
+    int sector = Shower->sector();
+    if (not Shower->isValid())
```
Comment (perrotta) [inline DQM/L1TMonitor/src/L1TStage2Shower.cc:171]: This check should be the first instruction in the loop

File: DQM/L1TMonitor/src/L1TStage2Shower.cc [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+          cscShowerChamber->Fill(chamber, 2.5);
+        if (cscShower->isTightInTime())
+          cscShowerChamber->Fill(chamber, 3.5);
+      }
+    }
+  }
+
+  // Fill EMTF regional shower plots
+  for (auto Shower = EmtfShowers->begin(); Shower != EmtfShowers->end(); ++Shower) {
+    int endcap = Shower->endcap();
```
Comment (perrotta) [inline DQM/L1TMonitor/src/L1TStage2Shower.cc:169]: `endcap` and `sector` can be better defined inside the `if` block that starts at line 173

File: DQM/L1TMonitor/src/L1TStage2Shower.cc [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  emtfShowerTypeOccupancy->setBinLabel(2, "ME- 2Loose", 2);
+  emtfShowerTypeOccupancy->setBinLabel(1, "ME- Tight", 2);
+}
+
+void L1TStage2Shower::analyze(const edm::Event& e, const edm::EventSetup& c) {
+  edm::Handle<l1t::RegionalMuonShowerBxCollection> EmtfShowers;
+  e.getByToken(EMTFShowerToken, EmtfShowers);
+
+  edm::Handle<CSCShowerDigiCollection> CscShowers;
+  e.getByToken(CSCShowerToken, CscShowers);
```
Comment (perrotta) [inline DQM/L1TMonitor/src/L1TStage2Shower.cc:83]: If you don't need the Handle's then it is faster:
```suggestion
  l1t::RegionalMuonShowerBxCollection const& EmtfShowers = e.get(EMTFShowerToken);
  CSCShowerDigiCollection const& CscShowers = e.get(CSCShowerToken);
```

**Review by perrotta — COMMENTED** [review]
(Please check that I did not add typos in the possible suggestions I'm writing here)

**Comment by zuoxunwu** [comment]
@perrotta Thanks for the suggestions and sorry for the late reply.
I have applied the changes following your suggestion.
Just as a validation, I tested it locally and everything worked fine.
This PR has been updated with the new commit.

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**